src/server/pkg/storage/gc/client.go

```diff
@@ -0,0 +1,265 @@
+	Ctx() context.Context
+	DeleteChunks([]chunk.Chunk) error
+	FlushDeletes([]chunk.Chunk) error
+}
+
+type Client struct {
+	server Server
+	db     *sql.DB
+}
+
+type Reference struct {
+	sourcetype string
+	source     string
+	chunk      chunk.Chunk
+}
+
+func initializeDb(db *sql.DB, ctx context.Context) error {
+	// TODO: move initialization somewhere more consistent
+	_, err := db.ExecContext(ctx, `
+do $$ begin
+ create type reftype as enum ('chunk', 'job', 'semantic');
+exception
+ when duplicate_object then null;
+end $$
+  `)
+	if err != nil {
+		return err
+	}
+
+	_, err = db.ExecContext(ctx, `
+create table if not exists refs (
+ sourcetype reftype not null,
+ source text not null,
+ chunk text not null,
+ primary key(sourcetype, source, chunk)
+)`)
+	if err != nil {
+		return err
+	}
+
+	_, err = db.ExecContext(ctx, `
+create table if not exists chunks (
+ chunk text primary key,
+ deleting timestamp
+)`)
+	if err != nil {
+		return err
+	}
+
+	_, err = db.ExecContext(ctx, `create index if not exists idx_chunk on refs (chunk)`)
+	if err != nil {
+		return err
+	}
+
+	_, err = db.ExecContext(ctx, `create index if not exists idx_sourcetype_source on refs (sourcetype, source)`)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func NewClient(server Server, host string, port int16) (*Client, error) {
+	connStr := fmt.Sprintf("host=%s port=%d dbname=pgc user=pachyderm password=elephantastic sslmode=disable", host, port)
+	connector, err := pq.NewConnector(connStr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Opening a connection is done lazily, statement preparation will connect
+	db := sql.OpenDB(connector)
+	ctx := server.Ctx()
+
+	err = initializeDb(db, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		db:     db,
+		server: server,
+	}, nil
+}
+
+func readChunksFromCursor(cursor *sql.Rows) []chunk.Chunk {
+	chunks := []chunk.Chunk{}
+	for cursor.Next() {
+		var hash string
+		if err := cursor.Scan(&hash); err != nil {
+			return nil
+		}
+		chunks = append(chunks, chunk.Chunk{Hash: hash})
+	}
+	return chunks
+}
+
+func (gcc *Client) ReserveChunks(job string, chunks []chunk.Chunk) error {
+	if len(chunks) == 0 {
+		return nil
+	}
+
+	ctx := gcc.server.Ctx()
+
+	// TODO: check for conflict errors and retry in a loop
+	txn, err := gcc.db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelSerializable})
+	if err != nil {
+		return err
+	}
+
+	questions := []string{}
+	for i := 0; i < len(chunks); i++ {
+		questions = append(questions, fmt.Sprintf("($%d)", 2+i))
+	}
+
+	query := `
+with
+added_chunks as (
+ insert into chunks (chunk)
+  values ` + strings.Join(questions, ",") + `
+ on conflict (chunk) do update set chunk = excluded.chunk
+ returning chunk, deleting
+),
+added_refs as (
+ insert into refs (chunk, source, sourcetype)
+  select
+   chunk,
+   ($1) as source,
+   'job'::reftype as sourcetype
+  from added_chunks where
+   deleting is null
+)
+
+select chunk from added_chunks where deleting is not null;
```

Comment:
Are you planning to use an ORM eventually? I feel like embedded query strings is probably going to become unmaintainable over time.